californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -397,35 +477,41 @@ public void sendResponse(final Exchange exchange, final Response response) {
 	@Override
 	public void receiveResponse(final Exchange exchange, final Response response) {
 
-		if (exchange.getRequest().isCanceled()) {
-			// do not continue fetching blocks if canceled
-			// reject (in particular for Block+Observe)
-			if (response.getType()!=Type.ACK) {
-				LOGGER.finer("rejecting blockwise transfer for canceled Exchange");
-				EmptyMessage rst = EmptyMessage.newRST(response);
-				sendEmptyMessage(exchange, rst);
-				// Matcher sets exchange as complete when RST is sent
-			}
+		if (isTransparentBlockwiseHandlingEnabled()) {
 
-		} else if (!response.hasBlockOption()) {
+			if (exchange.getRequest().isCanceled()) {
```

Comment:
if (isTransparentBlockwiseHandlingEnabled()) {

			if (exchange.getRequest().isCanceled()) {
				// do not continue fetching blocks if canceled
				// reject (in particular for Block+Observe)
				if (response.getType() != Type.ACK) {
					LOGGER.finer("rejecting blockwise transfer for canceled Exchange");
					EmptyMessage rst = EmptyMessage.newRST(response);
					sendEmptyMessage(exchange, rst);
					// Matcher sets exchange as complete when RST is sent
				}

I think, the 
			if (exchange.getRequest().isCanceled()) {
				// do not continue fetching blocks if canceled

is at the right place here. But

				if (response.getType() != Type.ACK) {
					LOGGER.finer("rejecting blockwise transfer for canceled Exchange");
					EmptyMessage rst = EmptyMessage.newRST(response);
					sendEmptyMessage(exchange, rst);
					// Matcher sets exchange as complete when RST is sent
				}

is not only at the wrong place, it seems to be mixed up. I don't see, none ACK should be answered with RST.